psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -131,93 +131,42 @@ void DirectJK::preiterations() {
-
-            if (do_wK_) {
-                prev_wK_ao_.push_back(std::make_shared<Matrix>("wK Prev", wK_ao_[N]->nrow(), wK_ao_[N]->ncol()));
-                delta_wK_ao_.push_back(std::make_shared<Matrix>("Delta wK", wK_ao_[N]->nrow(), wK_ao_[N]->ncol()));
-            }
-                
-            if (do_J_) {
-                prev_J_ao_.push_back(std::make_shared<Matrix>("J Prev", J_ao_[N]->nrow(), J_ao_[N]->ncol()));
-                delta_J_ao_.push_back(std::make_shared<Matrix>("Delta J", J_ao_[N]->nrow(), J_ao_[N]->ncol()));
-            }
-        
-            if (do_K_) {
-                prev_K_ao_.push_back(std::make_shared<Matrix>("K Prev", K_ao_[N]->nrow(), K_ao_[N]->ncol()));
-                delta_K_ao_.push_back(std::make_shared<Matrix>("Delta K", K_ao_[N]->nrow(), K_ao_[N]->ncol()));
+    if (do_incfock_iter_) {
+        size_t njk = D_ao_.size();
+
+        // If there is no previous pseudo-density, this iteration is normal
+        if (initial_iteration_ || D_prev_.size() != njk) {
+	        initial_iteration_ = true;
+
+            D_ref_ = D_ao_;
+            zero();
+        } else { // Otherwise, the iteration is incremental
+            for (size_t jki = 0; jki < njk; jki++) {
+                D_ref_[jki] = D_ao_[jki]->clone();
+                D_ref_[jki]->subtract(D_prev_[jki]);
             }
         }
     } else {
-        for (size_t N = 0; N < D_ao_.size(); N++) {
-            delta_D_ao_[N]->copy(D_ao_[N]);
-            delta_D_ao_[N]->subtract(prev_D_ao_[N]);
-        }
+        D_ref_ = D_ao_;
+        zero();
     }
 }
-void DirectJK::incfock_postiter() {
-    if (do_incfock_iter_) {
-        for (size_t N = 0; N < D_ao_.size(); N++) {
-
-            if (do_wK_) {
-                prev_wK_ao_[N]->add(delta_wK_ao_[N]);
-                wK_ao_[N]->copy(prev_wK_ao_[N]);
-            }
-
-            if (do_J_) {
-                prev_J_ao_[N]->add(delta_J_ao_[N]);
-                J_ao_[N]->copy(prev_J_ao_[N]);
-            }
-
-            if (do_K_) {
-                prev_K_ao_[N]->add(delta_K_ao_[N]);
-                K_ao_[N]->copy(prev_K_ao_[N]);
-            }
 
-            prev_D_ao_[N]->copy(D_ao_[N]);
-        }
-    } else {
-        for (size_t N = 0; N < D_ao_.size(); N++) {
-            if (do_wK_) prev_wK_ao_[N]->copy(wK_ao_[N]);
-            if (do_J_) prev_J_ao_[N]->copy(J_ao_[N]);
-            if (do_K_) prev_K_ao_[N]->copy(K_ao_[N]);
-            prev_D_ao_[N]->copy(D_ao_[N]);
-        }
+void DirectJK::incfock_postiter() {
+    // Save a copy of the density for the next iteration
+    D_prev_.clear();
+    for(auto const &Di : D_ao_) {
+        D_prev_.push_back(Di->clone());
     }
 }
 
 void DirectJK::compute_JK() {
-
-    // zero out J, K, and wK matrices
-    zero();
-
+   
 #ifdef USING_BrianQC
     if (brianEnable) {
+        // zero out J, K, and wK matrices
+        zero();
```

Comment:
just checking that editing brianqc code intentional.

Reply:
Indeed it is!

The original DirectJK code has a call to zero() at the very beginning of compute_JK(). Unfortunately, this zero() call breaks the IncFock code when using the new IncFock formalism, so essentially, the call to zero() was moved down further in DirectJK. _However_, the call to zero() also occurs before the BrianQC code. So, I assumed that just flat-out getting rid of the call to zero() at the beginning would break the BrianQC code. The addition of zero() to BrianQC, then, is essentially making sure that zero() is still called at the beginning when BrianQC is used.

Thinking about it further, though, it's possible this same effect can be had by just replacing the original call to zero() with something like

`if (!incfock_) zero();`

Then, the zero() call is only done for non-IncFock runs without changing the BrianQC code at all. Might be safer to do it that way, since I am not sure if we have immediate access to BrianQC to test changes to BrianQC code.

Reply:
We don't have immediate BrianQC access, so I think the consolidated `zero()` you propose would be good.

Reply:
Will do, thanks! My only concern is some edge case where someone uses BrianQC while requesting IncFock, then the initial zeroing doesn't happen, and then BrianQC possibly breaks.